coredns-operator/pkg/controller/coredns/coredns_controller.go

```diff
@@ -0,0 +1,279 @@
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileCoreDNS) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling CoreDNS")
+
+	// Fetch the CoreDNS instance
+	instance := &addonsv1alpha1.CoreDNS{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if instance.Spec.Corefile == "" {
+		instance.Spec.Corefile = DefaultCorefile
+		reqLogger.Info("Updating CoreDNS with default Corefile")
+		err = r.client.Update(context.TODO(), instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	foundDeployment := &appsv1.Deployment{}
+	err = r.createOrFetch(reqLogger, instance, newDeploymentForCR(instance), foundDeployment)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
+
+	service := newServiceForCR(instance)
+	dnsIP, err := r.calculateDNSClusterIP()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if instance.Spec.ClusterDNS {
+		service.Spec.ClusterIP = dnsIP
+	}
+	foundService := &corev1.Service{}
+	err = r.createOrFetch(reqLogger, instance, service, foundService)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	} else if err == nil {
+		// The object was found already existing in the cluster
+		// Check that the Service's ClusterIP is correct -- if not, update it
+		if instance.Spec.ClusterDNS {
+			if foundService.Spec.ClusterIP != dnsIP {
+				reqLogger.Info("Re-creating Service", foundService.Namespace, "/", foundService.Name, "with operator-managed ClusterIP:", dnsIP)
+				// ClusterIP field is immutable -- delete it, next reconcile will re-create
+				err = r.client.Delete(context.TODO(), foundService)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+			}
+		} else {
+			if foundService.Spec.ClusterIP == dnsIP {
+				reqLogger.Info("Re-creating Service", foundService.Namespace, "/", foundService.Name, "-- removing operator-managed ClusterIP:", dnsIP)
+				// ClusterIP field is immutable -- delete it, next reconcile will re-create
+				err = r.client.Delete(context.TODO(), foundService)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+			}
+		}
+	}
+
+	foundConfigMap := &corev1.ConfigMap{}
+	err = r.createOrFetch(reqLogger, instance, newConfigMapForCR(instance), foundConfigMap)
```

Comment:
Do we have to do this?  I'm obviously biased, but I much prefer the declarative approach both from the point of view of not requiring recompilation, and from the point of view of creating something that we can realistically expect everyone to use.

Could we combine operator-sdk with the [declarative library](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern) ?  What additional things does operator-sdk add over kubebuilder if we do that?